core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java

```diff
@@ -0,0 +1,361 @@
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return source.isEmpty();
+    }
+
+    @Override
+    public long demand() {
+        return source.demand();
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return source.whenComplete();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super U> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        source.subscribe(
+                new FlatMapAggregatingSubscriber<>(subscriber, function, executor, maxConcurrency),
+                executor, options);
+    }
+
+    @Override
+    public void abort() {
+        source.abort();
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        source.abort(cause);
+    }
+
+    private static final class FlatMapAggregatingSubscriber<T, U> implements Subscriber<T>, Subscription {
+        private final int maxConcurrency;
+
+        private final Subscriber<? super U> downstream;
+        private final Function<T, StreamMessage<U>> function;
+        private final EventExecutor executor;
+        private final Set<FlatMapSubscriber<T, U>> sourceSubscriptions;
+        private final Queue<U> buffer;
+
+        @Nullable
+        private volatile Subscription upstream;
+        private volatile boolean canceled;
+
+        private long requestedByDownstream;
+        private int pendingSubscriptions;
+        private boolean completing;
+
+        FlatMapAggregatingSubscriber(Subscriber<? super U> downstream,
+                                     Function<T, StreamMessage<U>> function,
+                                     EventExecutor executor,
+                                     int maxConcurrency) {
+            requireNonNull(downstream, "downstream");
+            requireNonNull(function, "function");
+            requireNonNull(executor, "executor");
+
+            this.downstream = downstream;
+            this.function = function;
+            this.executor = executor;
+            this.maxConcurrency = maxConcurrency;
+
+            sourceSubscriptions = new HashSet<>();
+            buffer = new LinkedList<>();
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+
+            upstream = subscription;
+            downstream.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+
+            if (canceled) {
+                StreamMessageUtil.closeOrAbort(item);
+                return;
+            }
+
+            final StreamMessage<U> newStreamMessage = function.apply(item);
+            newStreamMessage.subscribe(new FlatMapSubscriber<>(this));
```

Comment:
Is it OK to just pass executor here? I don't think there are any options that need to be applied

Reply:
@ikhoon Gentle ping

Reply:
I tried passing the executor here and I'm running into an issue I'm not sure how to address. 

This code calls `subscribe`, which calls `onSubscribe` on `FlatMapSubscriber`. That will call `subscribeChild` on `FlatMapAggregatingSubscriber`, which will add the subscriber to `sourceSubscriptions` and then call `request` on `FlatMapSubscriber`. Then, if `newStreamMessage` is already completed it will call `onComplete`, which leads to the subscription being removed from `sourceSubscriptions`. 

Essentially, in one call chain we go from `onSubscribe` to `sub.request(1)` all the way to `onComplete`, but since that removes the subscription from `sourceSubscriptions` I get a `ConcurrentModificationException` since the call to `sub.request(1)` is done in a `forEach`. 

I guess I should just collect the subs into a new collection before calling `forEach` to avoid that? I'm not sure if there's a better approach

Reply:
I see. How about always using `executor` to remove the child?
`executor.execute(() -> handleCompleteChild(child));`

Reply:
If that works, we can remove all `eventLoop.inEventLoop()` check because we only use the executor.